cmd/kubeapps-apis/plugins/kapp_controller/packages/v1alpha1/server_data_adapters.go

```diff
@@ -308,3 +312,75 @@ func (s *Server) getInstalledPackageDetail(pkgInstall *packagingv1alpha1.Package
 
 	return installedPackageDetail, nil
 }
+
+func (s *Server) newSecret(installedPackageName, values, targetNamespace string) (*k8scorev1.Secret, error) {
+	return &k8scorev1.Secret{
+		TypeMeta: metav1.TypeMeta{
+			Kind:       k8scorev1.ResourceSecrets.String(),
+			APIVersion: k8scorev1.SchemeGroupVersion.WithResource(k8scorev1.ResourceSecrets.String()).String(),
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			// TODO(agamez): think about name collisions
+			Name:      fmt.Sprintf("%s-values", installedPackageName),
+			Namespace: targetNamespace,
+		},
+		Data: map[string][]byte{
+			// TODO(agamez): check the actual value for the key.
+			// Assuming "values.yaml" perhaps is not always true.
+			// Perhaos this info is in the "package" object?
+			"values.yaml": []byte(values),
+		},
+		Type: "Opaque",
+	}, nil
+}
+
+func (s *Server) newPkgInstall(installedPackageName, targetNamespace, packageRefName, pkgVersion string, reconciliationOptions *corev1.ReconciliationOptions) (*packagingv1alpha1.PackageInstall, error) {
+	pkgInstall := &packagingv1alpha1.PackageInstall{
+		TypeMeta: metav1.TypeMeta{
+			Kind:       pkgInstallResource,
+			APIVersion: fmt.Sprintf("%s/%s", packagingv1alpha1.SchemeGroupVersion.Group, packagingv1alpha1.SchemeGroupVersion.Version),
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      installedPackageName,
+			Namespace: targetNamespace,
+		},
+		Spec: packagingv1alpha1.PackageInstallSpec{
+			// TODO(agamez): remove this when we have a real service account in the UI
+			ServiceAccountName: "default",
+			// ServiceAccountName: "default-ns-sa",
+			// TODO(agamez): check what this "Cluster" is for
+			// Cluster: &kappctrlv1alpha1.AppCluster{
+			// 	Namespace:           targetNamespace,
+			// 	KubeconfigSecretRef: &kappctrlv1alpha1.AppClusterKubeconfigSecretRef{},
+			// },
```

Comment:
Yeah, it's Carvel's way of supporting deployments across clusters (without having kapp-controller on those other clusters). I've chatted with the carvel team about this. My concern is that with this you're relying on secrets on the cluster that give very permissive RBAC to another cluster. Their response was that this is how others (kubernetes dashboard) do things, and that the secret is no worse than a service account for this cluster (which yes, it may be in some cases limited to a specific namespace etc., but it's a level up from RBAC imo).

As it is, with our UX we'd be assuming kapp-controller on the other cluster and talking to it with the users creds as normal. I don't think we have to support this case of deploying to another cluster without kapp-controller being installed there, at least yet.

Reply:
Thanks for the context here, really useful!! When I saw it for the very first time I thought -oh, out-of-the-box multicluster! - but then I noticed it wasn't what I expected, haha.
I've added a note in the code pointing to this comment.